projects/spectator/src/lib/internals/element-focus.ts

```diff
@@ -1,13 +1,69 @@
 import { dispatchFakeEvent } from '../dispatch-events';
+import { isRunningInJsDom } from '../utils';
+
+/** Property name added to HTML Elements to ensure we don't double-patch focus methods on an element. */
+const IS_FOCUS_PATCHED_PROP = '_patched_focus';
+
+/** Ensures that a single set of matching focus and blur events occur when HTMLElement.focus() is called. */
+class FocusEventWatcher implements EventListenerObject {
+
+  private readonly _active: Element | null;
+  private _blurred = false;
+  private _focused = false;
+
+  constructor(private readonly _e: HTMLElement) {
+    this._active = _e.ownerDocument.activeElement;
+    this._e.addEventListener('focus', this);
+    this._active?.addEventListener('blur', this);
+  }
+  public handleEvent(evt: Event): void {
+    if (evt.type === 'focus') {
+      this._focused = true;
+    }
+    else if (evt.type === 'blur') {
+      this._blurred = true;
+    }
+  }
+
+  /**
+   * If focus and blur events haven't occurred, fire fake ones.
+   */
+  public ensureFocusEvents() {
+    this._e.removeEventListener('focus', this);
+    this._active?.removeEventListener('blur', this);
+
+    // Ensure activeElement is blurred
+    if (this._active && !this._blurred && this._active === this._e.ownerDocument.activeElement) {
```

Comment:
Isn't the following condition more accurate? 
```ts
if(!this._blurred && this.activeElement === this.element) {}
```

Reply:
This check is to make sure that the `document.activeElement` still matches the `document.activeElement` that was recorded when the watcher was created (before the browser's `element.focus()` was called). It could conceivably change, and I don't think we'd want the initial `activeElement` to receive a blur event if the `activeElement` has since changed.